plugins/kapt3/kapt3-base/src/org/jetbrains/kotlin/kapt3/base/KaptContext.kt

```diff
@@ -48,7 +48,9 @@ open class KaptContext(val options: KaptOptions, val withJdk: Boolean, val logge
                 KaptJavaLog17(
                     options.projectBaseDir, newContext,
                     interceptorData, options[KaptFlag.MAP_DIAGNOSTIC_LOCATIONS]
-                )
+                ).apply {
```

Comment:
The code above seems complex. I wonder why it is like that. It would seem a lot simpler to me to just pass in the three writers in the constructor and have an `init` block in `KaptJavaLog17` that sets up the writers the way we want them:

```
     init {
        setWriter(WriterKind.ERROR, errorWriter)
        setWriter(WriterKind.WARNING, warnWriter)
        setWriter(WriterKind.NOTICE, infoWriter)
        setWriter(WriterKind.STDERR, errorWriter)
        setWriter(WriterKind.STDOUT, infoWriter)
    }
```

Is there a deeper reason for the reflection above that I do not understand? I have no experience with the javac API, so maybe I'm way off?

Reply:
Also, we should figure out if there are tests where we could have caught this change from before JDK17.

Reply:
Yeah, using `setWriter` seems like a better approach. And it will allow to remove `KaptJavaLog17` altogether.

However, it looks like setting `errWriter` to `WriterKind.STDERR` makes the build fail because some javac messages are reported as errors (see Daniel's repo from KT-54030). Also, `WriterKind.STDERR/STDOUT` are not accessible on JDK 8. I guess we shouldn't set writers for `STDERR` and `STDOUT`, so that those messages would be dumped simply as text to stderr/stdout correspondingly, as they do with `KaptJavaLog`.

Reply:
For tests, I think we should use this: https://github.com/JetBrains/kotlin/blob/2c889f56fad8ce6eb496ec658a7e20efe20f09dd/libraries/tools/kotlin-gradle-plugin-integration-tests/src/test/kotlin/org/jetbrains/kotlin/gradle/Kapt3IT.kt#L129

Reply:
Thanks for the guidance! I admit I went with the simplest fix, but I also agree we can cleanup `KaptJavaLog17` and use `setWriter`. I updated the commit, please take a look.

As for testing, `Kapt3IT` didn't work for me since I couldn't validate the issue by only asserting on a test project's Gradle output, however a more 'lower level' test such as `AbstractKotlinKapt3IntegrationTestBase` worked well since I could repro the issue by checking a `MessageCollector`'s reported severity.